File: 1408-find-the-smallest-divisor-given-a-threshold/find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int Maxelement(vector<int> &nums){
        int ele=0;
        for(int i=0;i<nums.size();i++){
            ele=max(ele,nums[i]);
        }
        return ele;
    }
    int smallestDivisor(vector<int>& nums, int threshold) {
        int low=1;
        int high=Maxelement(nums);
        int sum=0;
        int n=nums.size();
        int ans=-1;
        while(low<=high){
            int mid=(low+high)/2;
            sum=0;
            for(int i=0;i<n;i++){
                sum += ceil((double)nums[i] / mid);
            }
            if(sum>threshold){
                low=mid+1;
            }
            else if(sum<=threshold){
                high=mid-1;
                ans=mid;
            }
        }
        return ans;
    }
};

File: 1408-find-the-smallest-divisor-given-a-threshold/FindTheSmallestDivisorGivenAThreshold.java
class Solution {
    static boolean isValid(int divisor, int[] nums, int threshold) {
        int sum = 0;
        for (int num : nums) {
            sum += (num + divisor - 1) / divisor;  // Efficient ceil
            if (sum > threshold) return false;
        }
        return true;
    }

    public int smallestDivisor(int[] nums, int threshold) {
        int low = 1, high = Arrays.stream(nums).max().getAsInt();
        int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isValid(mid, nums, threshold)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }
}
